### ISSUE TITLE
filter2 fails with boundry='replicate' and larger then 3x3 filter

### ISSUE THREAD

**Issue by ghaarsma** [aoles/EBImage#8]
```R
x <- matrix(runif(7*13),7,13)

k <- matrix(1,nrow = 5,ncol = 5);k<-k/sum(k)

x1 <- filter2(x,k,boundary = 'circular')  # This works fine
x2 <- filter2(x,k,boundary = 'replicate') # This has an error
```

Error in abind(x, duplicate(x, d[1L], 1:d[2L], cf[1L], 1L), duplicate(x,  : 
  arg 'X2' has dims=1, 26; but need dims=X, 13

There is also a problem with the section:
```R
  if ( boundary!="circular" ) {
    y = asub(y, list(1:d[1L], 1:d[2L]), 1:2)
  }
```
It does not cut out the center image, but top left corner, since indices start at 1 in the list. I have reworked the `replicate` boundary code a bit to get it to work.  Let me know if the code below makes sense and you can use it and please check it on other objects to make sure it works.
```R
# 2D convolution-based linear filter for images and matrix data

# Copyright (c) 2007-2015, Andrzej Oleś, Gregoire Pau, Oleg Sklyar

# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public License
# as published by the Free Software Foundation; either version 2.1
# of the License, or (at your option) any later version.

# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

# See the GNU Lesser General Public License for more details.
# LGPL license wording: http://www.gnu.org/licenses/lgpl.html

## - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

filter2_fix = function(x, filter, boundary = c("circular", "replicate")) {
  if ( is.numeric(boundary) ) {
    val = boundary[1L]
    boundary = "linear"
  }
  else
    boundary = match.arg(boundary)
  
  validObject(x)
  validObject(filter)
  
  d = dx = dim(x)
  df = dim(filter)
  dnames = dimnames(x)
  
  if (any(df%%2==0)) stop("dimensions of 'filter' matrix must be odd")
  if (any(dx[1:2]<df)) stop("dimensions of 'x' must be bigger than 'filter'")
  
  cf = df%/%2
  
  switch(boundary,
         ## default mode just wraps around edges
         circular = {},
         ## pad with a given value
         linear = {
           dx[1:2] = dx[1:2] + cf[1:2]
           xpad = array(val, dx)
           ## is there a better way of doing this?
           imageData(x) = do.call("[<-", c(quote(xpad), lapply(d, function(x) enquote(1:x)), quote(x)) )
         },
         replicate = {
           dx[1:2] = dx[1:2] + df[1:2] - 1L

           rep.dim <- function(x,dim,index,times) {
             x.s <- asub(x,idx=index,dims = dim,drop = FALSE)
             dim.s <- dim(x.s)
             dim.s[dim] <- times
             if (dim==2) {
               x.s <- split(x.s,ceiling(seq_along(x.s)/nrow(x)))
               x.s <- sapply(x.s,FUN = rep,times=times)
             } else {
               x.s <- rep(x.s,each=times)
             }
             out <- array(x.s,dim = dim.s)
           }
           
           # Add left and right colums
           lc <- rep.dim(x,2,1,cf[2])
           rc <- rep.dim(x,2,ncol(x),cf[2])
           x <- abind(lc,x,rc,along = 2)
           # Add top and bottom rows
           tr <- rep.dim(x,1,1,cf[1])
           br <- rep.dim(x,1,nrow(x),cf[1])
           imageData(x) <- abind(tr,x,br,along = 1)
         }
  )
  
  ## create fft filter matrix
  wf = matrix(0.0, nrow=dx[1L], ncol=dx[2L])
  
  wf[c(if (cf[1L]>0L) (dx[1L]-cf[1L]+1L):dx[1L] else NULL, 1L:(cf[1L]+1L)), 
     c(if (cf[2L]>0L) (dx[2L]-cf[2L]+1L):dx[2L] else NULL, 1L:(cf[2L]+1L))] = filter
  
  wf = fftw2d(wf)
  
  pdx = prod(dx[1:2])
  
  .filter = function(xx) Re(fftw2d(fftw2d(xx)*wf, inverse=1)/pdx)
  
  ## convert to a frame-based 3D array
  if ( length(dx) > 2L ) {    
    y = apply(x, 3:length(dx), .filter)
    dim(y) = dx
  }
  else {
    y = .filter(x)
  }
  
  if ( boundary!="circular" ) {
    y = asub(y, list(cf[1]+(1:d[1L]), cf[2]+(1:d[2L])), 1:2)
  }
  
  dimnames(y) = dnames
  
  imageData(x) <- y
  x
}
```

**Comment by aoles** [comment]
Thank you very much for your detailed report and for the suggested fix.
I will get back to you in the next few days once I review your code.

Cheers,
Andrzej

**Comment by aoles** [comment]
Thank you again for looking into the problem and for taking the lead in fixing the bug instead of just complaining :)

The lines
```r
if ( boundary!="circular" ) {
    y = asub(y, list(1:d[1L], 1:d[2L]), 1:2)
  }
```
are actually correct. The relevant image data starts in the top left corner because in the "replicate" and "linear" boundary modes the canvas is extended only towards the right and the bottom. This approach reduces slightly the bookkeeping compared to the traditional one where the extension is added symmetrically around the original image, and I prefer to keep it this way.

Thank you for your suggested solution, your code is much cleaner than my original implementation. I modified it accounting for the above, and also tweaked a little bit some parts, mostly formatting.

Cheers,
Andrzej

**Comment by ghaarsma** [comment]
Hi Andrzej,

Please reconsider the boundary issues with "replicate" and "linear" modes. It should really take the centerpoint of the extended canvas. My original pull request did not address this for the linear case. I have create a new pull request, addressing this issue. As an added bonus you can now add individual values for each frame (for example if you want to match the color in a RGB image). The pull request also has updated documentation.

Please see this trivial filter example why this is important. With this filter you should just get the original image back, no matter what. This works for circular but with your implementation fails for linear and replicate boundaries.
```R
library(EBImage)

x=matrix(1:32,nrow = 4,ncol = 8)
kernel <- matrix(0,nrow = 3,ncol = 5);kernel[2,3]<-1;x

filter2(x,kernel,boundary = 'circular')
filter2(x,kernel,boundary = 'replicate')
filter2(x,kernel,boundary = 50)
```

Also, see this for more background info: http://www.coe.utah.edu/~cs4640/slides/Lecture5.pdf (page 16-20)

**Comment by aoles** [comment]
Hi Gabe,

thank you for following up on this! I'm not really convinced that the problems you experience are related to the way the padding is added to the image. According to the slides which you refer to convolution is shift-invariant (Slide 28), so in principle it shouldn't matter whether I add the padding symmetrically and cut out the center part, or extend the canvas only at the right/bottom and take the upper left part afterwards. In fact, when I run your examples on the current EBImage version and check for object identity I get:

```r
library(EBImage)

x = matrix(1:32, nrow = 4, ncol = 8)
kernel = matrix(0, nrow = 3, ncol = 5)
kernel[2,3] = 1

fc = filter2(x, kernel, boundary = 'circular')
fr = filter2(x, kernel, boundary = 'replicate')
fl = filter2(x, kernel, boundary = 50)

storage.mode(x) = "double"

identical(x, fc)
## [1] TRUE
identical(x, fr)
## [1] TRUE
identical(x, fl)
## [1] FALSE
```
So at least for `boundary="replicate"` my approach seems fine. Re-running the above against your code installed through `devtools::install_github("ghaarsma/EBImage")`results in:

```r
identical(x, fc)
## [1] TRUE
identical(x, fr)
## [1] FALSE
identical(x, fl)
## [1] FALSE
```
So apparently your pull request doesn't solve the issue. As a downside, it breaks the replicate mode because you didn't account for the non-symmetric padding I introduced.

```r
fr
##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8]
## [1,]   10   14   18   22   26   30   30   30
## [2,]   11   15   19   23   27   31   31   31
## [3,]   12   16   20   24   28   32   32   32
## [4,]   12   16   20   24   28   32   32   32
```

I had a closer look at the differences between `x` and `fl` computed in my original code. It turns out that these are at the machine precision level:

```r
all.equal(x, fl)
## [1] TRUE
```
In conclusion, I'm not sure whether there is any actual problem with *EBImage* code, or the observed discrepancies are just the intrinsic nature of the FFT filtering. I will need to investigate this further.

Cheers,
Andrzej

**Comment by ghaarsma** [comment]
Hi Andrzej,

Yes my previous pull request was incorrect (I missed several of your edits when manually trying to merge our branches).

Yes we should be using all.equal to avoid machine precision levels.

Yes, after careful review on my end I have concluded that your approach with non-symmetric padding (for both linear and replicate) should work and is the more computational elegant solution.

Please see my latest pull request with a small modification on the linear boundary. This will make selecting the linear boundary implementation correct. It also allows you to use different boundary values for each frame. Handy for RGB images if you want to match color.

```R
library(EBImage)

x=matrix(as.numeric(1:32),nrow = 4,ncol = 8)
kernel <- matrix(0,nrow = 3,ncol = 5);kernel[2,3]<-1;x

fc <- filter2(x,kernel,boundary = 'circular')
fr <- filter2(x,kernel,boundary = 'replicate')
fb <- filter2(x,kernel,boundary = 50)

all.equal(fc,x)
all.equal(fr,x)
all.equal(fb,x)
```
Results in:
```R
> all.equal(fc,x)
[1] TRUE
> all.equal(fr,x)
[1] TRUE
> all.equal(fb,x)
[1] TRUE
```

**Comment by aoles** [comment]
Hi Gabe,

thank you for getting back to me with the suggested improvements to the linear boundary case. I will be happy to merge your changes once you address my comments to your pull request.

Cheers,
Andrzej
